pkg/analysis_server/lib/src/computer/computer_closingLabels.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analysis_server/protocol/protocol_generated.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/src/generated/source.dart';
+
+/**
+ * A computer for [CompilationUnit] closing labels.
+ */
+class DartUnitClosingLabelsComputer {
+  final LineInfo _lineInfo;
+  final CompilationUnit _unit;
+  final List<ClosingLabel> _closingLabels = <ClosingLabel>[];
+
+  DartUnitClosingLabelsComputer(this._lineInfo, this._unit);
+
+  /**
+   * Returns a list of closing labels, not `null`.
+   */
+  List<ClosingLabel> compute() {
+    _unit.accept(new _DartUnitClosingLabelsComputerVisitor(this));
+    return _closingLabels;
+  }
+}
+
+/**
+ * An AST visitor for [DartUnitClosingLabelsComputer].
+ */
+class _DartUnitClosingLabelsComputerVisitor
+    extends RecursiveAstVisitor<Object> {
+  final DartUnitClosingLabelsComputer computer;
+
+  _DartUnitClosingLabelsComputerVisitor(this.computer);
+
+  @override
+  Object visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (_spansManyLines(node)) {
+      // TODO(dantup) Is it right to read constructorName to get nameless constructors types?
+      final name = node.constructorName.type.name;
+      final label = name is PrefixedIdentifier ? name.prefix.name : name.name;
```

Comment:
Ok, I see why I'm confused.. For `a.Future.delayed`, the `constructorName.type` seems to be `a.Future` (sounds reasonable) and `constructorName.name` is `delayed` (also seems reasonable).

However, for `Class.fromThing`, `constructorName.type` seems to include `fromThing` and `constructorName.name` is null. This isn't what I'd expect; I'd expect `constructorName.type` to be `Class` and `constructorName.name` to be `fromThing`.

<img width="301" alt="screen shot 2017-08-09 at 18 09 13" src="https://user-images.githubusercontent.com/1078012/29134209-e07552fa-7d2d-11e7-92b2-92a19cd40cf7.png">

What's the reason for this (or what am I getting confused by)?

Btw - doing `node.constructorName.toString()` seems to give a reasonable output here, but it feels slightly weird to be calling `toString` on the AST node for this. WDYT?